internal/pally/registry.go

```diff
@@ -0,0 +1,272 @@
+// MustGauge constructs a new Gauge. It panics if it encounters an error.
+func (r *Registry) MustGauge(opts Opts) Gauge {
+	g, err := r.NewGauge(opts)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create Gauge with options %+v: %v", opts, err))
+	}
+	return g
+}
+
+// NewCounterVector constructs a new CounterVector.
+func (r *Registry) NewCounterVector(opts Opts) (CounterVector, error) {
+	opts = r.addConstLabels(opts)
+	if err := opts.validateVector(); err != nil {
+		return nil, err
+	}
+	v := newCounterVector(opts)
+	if err := r.register(v); err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// MustCounterVector constructs a new CounterVector. It panics if it encounters
+// an error.
+func (r *Registry) MustCounterVector(opts Opts) CounterVector {
+	v, err := r.NewCounterVector(opts)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create CounterVector with options %+v: %v", opts, err))
+	}
+	return v
+}
+
+// NewGaugeVector constructs a new CounterVector.
+func (r *Registry) NewGaugeVector(opts Opts) (GaugeVector, error) {
+	opts = r.addConstLabels(opts)
+	if err := opts.validateVector(); err != nil {
+		return nil, err
+	}
+	v := newGaugeVector(opts)
+	if err := r.register(v); err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// MustGaugeVector constructs a new GaugeVector. It panics if it encounters an
+// error.
+func (r *Registry) MustGaugeVector(opts Opts) GaugeVector {
+	v, err := r.NewGaugeVector(opts)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create GaugeVector with options %+v: %v", opts, err))
+	}
+	return v
+}
+
+// ServeHTTP implements http.Handler.
+func (r *Registry) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	r.handler.ServeHTTP(w, req)
+}
+
+func (r *Registry) register(m metric) error {
+	r.metricsMu.Lock()
+	r.metrics = append(r.metrics, m)
+	r.metricsMu.Unlock()
+
+	if err := r.prom.Register(m); err != nil {
+		return err
+	}
+	for _, fed := range r.federated {
+		if err := fed.Register(m); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Registry) addConstLabels(opts Opts) Opts {
+	if len(r.constLabels) == 0 {
+		return opts
+	}
+	labels := opts.copyLabels()
+	for k, v := range r.constLabels {
+		labels[k] = v
+	}
+	opts.ConstLabels = labels
+	return opts
+}
+
+type pusher struct {
+	reg     *Registry
+	stop    chan struct{}
+	stopped chan struct{}
+	scope   tally.Scope
+	ticker  *time.Ticker
+}
+
+func newPusher(r *Registry, scope tally.Scope, tick time.Duration) *pusher {
+	return &pusher{
+		reg:     r,
+		stop:    make(chan struct{}),
+		stopped: make(chan struct{}),
+		scope:   scope,
+		ticker:  time.NewTicker(tick),
+	}
+}
+
+func (p *pusher) Start() {
+	defer close(p.stopped)
+	// When stopping, do one last export to catch any stragglers.
+	defer p.push()
+
+	for {
+		select {
+		case <-p.stop:
+			return
+		case <-p.ticker.C:
+			p.push()
+		}
+	}
+}
+
+func (p *pusher) Stop() {
+	p.ticker.Stop()
+	close(p.stop)
+	<-p.stopped
+}
+
+func (p *pusher) push() {
```

Comment:
This logic should probably live inside of the `Registry`, and the pusher should just call `push(p.scope)` on the registry. prefer if types didn't reach into other types' private fields (especially when it requires locks etc)